File: tests/test_dssat_model_e2e.py
"""One-point real DSSAT smoke test for machines with a verified installation."""

import csv
import os
import platform
import shutil
import subprocess
from pathlib import Path

import pytest


pytestmark = [pytest.mark.integration, pytest.mark.slow]

ROOT = Path(__file__).resolve().parents[1]
FIXTURE = ROOT / "tests" / "fixtures" / "dssat_model_smoke"


def _find_dssat_executable():
    configured = os.getenv("DSSAT_EXE", "").strip()
    names = ["DSCSM048.EXE"] if platform.system() == "Windows" else ["dscsm048"]
    candidates = [Path(configured).expanduser()] if configured else []
    candidates.extend(ROOT.parent.joinpath("DSSAT48", name) for name in names)
    for name in names:
        found = shutil.which(name)
        if found:
            candidates.append(Path(found))
    for candidate in candidates:
        if candidate.is_file():
            return candidate.resolve()
    return None


def test_one_point_real_dssat_run(tmp_path):
    """Launch DSSAT and assert deterministic yield from the committed fixture."""
    executable = _find_dssat_executable()
    if executable is None:
        pytest.skip("verified DSSAT executable not installed; set DSSAT_EXE")

    profile_names = ("DSSATPRO.V48", "DSSATPRO.L48")
    if not any((executable.parent / name).is_file() for name in profile_names):
        pytest.skip(f"DSSAT support profile missing next to {executable}")

    run_dir = tmp_path / "point"
    shutil.copytree(FIXTURE, run_dir)
    result = subprocess.run(
        [str(executable), "B", "DSSBatch.V48"],
        cwd=run_dir,
        check=False,
        capture_output=True,
        text=True,
        timeout=60,
    )
    assert result.returncode == 0, (
        f"DSSAT exited {result.returncode}\nstdout:\n{result.stdout}\n"
        f"stderr:\n{result.stderr}"
    )

    summary = run_dir / "summary.csv"
    assert summary.is_file(), "DSSAT completed without summary.csv"
    with summary.open(newline="", encoding="utf-8-sig") as handle:
        rows = list(csv.DictReader(handle))
    assert len(rows) == 1
    assert rows[0]["CR"].strip() == "MZ"
    assert int(float(rows[0]["HWAM"])) == 6292
    assert int(float(rows[0]["CWAM"])) == 15971

File: tests/test_dssat_model_e2e.R
#!/usr/bin/env Rscript
# One-point real DSSAT smoke test for machines with a verified installation.

root <- normalizePath(getwd(), mustWork = TRUE)
fixture <- file.path(root, "tests", "fixtures", "dssat_model_smoke")

find_dssat_executable <- function() {
  configured <- trimws(Sys.getenv("DSSAT_EXE", ""))
  exe_name <- if (.Platform$OS.type == "windows") "DSCSM048.EXE" else "dscsm048"
  candidates <- character()
  if (nzchar(configured)) candidates <- c(candidates, path.expand(configured))
  candidates <- c(candidates, file.path(dirname(root), "DSSAT48", exe_name))
  on_path <- Sys.which(exe_name)
  if (nzchar(on_path)) candidates <- c(candidates, on_path)
  candidates <- unique(candidates)
  found <- candidates[file.exists(candidates)]
  if (length(found)) normalizePath(found[[1]], mustWork = TRUE) else ""
}

exe <- find_dssat_executable()
if (!nzchar(exe)) {
  message("SKIP: verified DSSAT executable not installed; set DSSAT_EXE")
  quit(status = 0)
}

profiles <- file.path(dirname(exe), c("DSSATPRO.V48", "DSSATPRO.L48"))
if (!any(file.exists(profiles))) {
  message("SKIP: DSSAT support profile missing next to ", exe)
  quit(status = 0)
}

work <- tempfile("dssat_model_smoke_")
dir.create(work, recursive = TRUE)
on.exit(unlink(work, recursive = TRUE, force = TRUE), add = TRUE)
copied <- file.copy(list.files(fixture, full.names = TRUE), work)
stopifnot(all(copied))

old_wd <- setwd(work)
on.exit(setwd(old_wd), add = TRUE)
output <- system2(exe, c("B", "DSSBatch.V48"), stdout = TRUE, stderr = TRUE)
status <- attr(output, "status")
if (is.null(status)) status <- 0L
if (status != 0L) {
  stop("DSSAT exited ", status, "\n", paste(output, collapse = "\n"))
}

if (!file.exists("summary.csv")) stop("DSSAT completed without summary.csv")
header <- strsplit(readLines("summary.csv", n = 1L), ",", fixed = TRUE)[[1]]
# DSSAT 4.8 writes one trailing field beyond the header in summary.csv.
summary <- utils::read.csv(
  "summary.csv",
  header = FALSE,
  skip = 1L,
  col.names = c(header, "EXTRA"),
  check.names = FALSE
)
stopifnot(nrow(summary) == 1L)
stopifnot(trimws(summary$CR[[1]]) == "MZ")
stopifnot(as.integer(summary$HWAM[[1]]) == 6292L)
stopifnot(as.integer(summary$CWAM[[1]]) == 15971L)
message("PASS: one-point real DSSAT model smoke test")
